### ISSUE TITLE
Travis CI Dev environment -> dev Lambda - Crawler

### ISSUE THREAD

**Issue by LyndonP** [ActoKids/devops#5]
Task is to configure Travis CI on dev environment of crawler team repository with automatic builds on pull request and set up an automatic deployment onto Lambda. 

Estimated time: 8 hrs
Actual time:      10 hrs

**Comment by dashinay** [comment]
Currently Travis CI is set up for continuous build on pull requests, though with some issues - crawler script is exiting with error. I am researching on deployment with Travis CI onto Lambda.
[How to deploy on lambda](https://docs.travis-ci.com/user/deployment/lambda/)

What preventing me from setting up an automated deployment is the issues with permission on IAM. For a deployment script I need access_key_id and secret_access_key. 
Also The AWS user that Travis deploys as must have the following IAM permissions in order to deploy:

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "ListExistingRolesAndPolicies",
            "Effect": "Allow",
            "Action": [
                "iam:ListRolePolicies",
                "iam:ListRoles"
            ],
            "Resource": "*"
        },
        {
            "Sid": "CreateAndListFunctions",
            "Effect": "Allow",
            "Action": [
                "lambda:CreateFunction",
                "lambda:ListFunctions"
            ],
            "Resource": "*"
        },
        {
            "Sid": "DeployCode",
            "Effect": "Allow",
            "Action": [
                "lambda:GetFunction",
                "lambda:UpdateFunctionCode",
                "lambda:UpdateFunctionConfiguration"
            ],
            "Resource": [
                "arn:aws:lambda:<region>:<account-id>:function:<name-of-function>"
            ]
        },
        {
           "Sid": "SetRole",
            "Effect": "Allow",
            "Action": [
                "iam:PassRole"
            ],
            "Resource": "arn:aws:iam::<account-id>:role/<name-of-role>"
        }
    ]
}

**Comment by dashinay** [comment]
I was able to get IAM permissions to the group resources and obtain access and secure keys. Was created IAM [role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html) for Lambda. Configured the deployment of crawler script into the lambda. The access keys have been placed into the travis environmental variables.

Configuration successfully build the script and was deployed onto aws lambda.
[Travis CI build](https://travis-ci.org/ActoKids/web-crawler/builds/489236401)
[Travis yaml file](https://github.com/ActoKids/web-crawler/blob/dev/.travis.yml)

![lambdafunc](https://user-images.githubusercontent.com/34629972/52752697-96a64100-2fa8-11e9-97ac-19ad363c0ee0.png)


Permissions 'AWSLambdaRole` and `AWSLambdaFullAccessRole` has been added to the DevOps group.

Resources used:
[Deploying aws lambda functions](https://dev.to/codevbus/deploy-aws-lambda-functions-with-aws-sam-cli-and-travis-ci-part-2-2goh)
[Building a python project with travis ci](https://docs.travis-ci.com/user/languages/python/)
[pull request](https://github.com/ActoKids/web-crawler/pull/5/commits/93d9971ba880643d95a595f033591d361ef72a36)

**Comment by toddysm** [comment]
Few things:
- We need a Wiki page explaining what the build does, what is the input and what the result should look like.
- Why do you run the crawler as part of the build? The goal is to deploy the Python code to the Lambda, not to run it
- you have `function_name: "ak-crawler-env-lbd"` in your YAML - how come this gets deployed to different Lambda?
- Did you test other people's implementations or have any issues filed to yours?

**Comment by dashinay** [comment]
1. I have added the [Wiki](https://github.com/ActoKids/web-crawler/wiki/Travis-CI)

2. Fixed an error by removing the build part of crawler and most of the dependencies, 
there is only deployment now.
https://github.com/ActoKids/web-crawler/pull/10

3. I had an old image attached and I missed to upload the changes and have provided a correct image of a the deployed Lambda function.

4. There was an security vulnerability issue in UI team dev branch. DevOps team had received a warning of security vulnerability with the message "jQuery before 3.0.0 is vulnerable to Cross-site Scripting (XSS) attacks when a cross-domain Ajax request is performed without the dataType option, causing text/javascript responses to be executed".
I was able to make a hot fix for by changing the version of JQuery to 3.0.0 or later.
PR : https://github.com/ActoKids/web-ui/pull/15
